lib/metrics/context.js

```
@@ -6,43 +6,178 @@
-
-  add: function (data, metadata, doNotTrack) {
-    if (metadata) {
-      data.time = Date.now()
-      data.flow_id = metadata.flowId
-      data.flow_time = calculateFlowTime(data.time, metadata.flowBeginTime)
-      data.context = metadata.context
-      data.entrypoint = metadata.entrypoint
-      data.migration = metadata.migration
-      data.service = metadata.service
-
-      if (! doNotTrack) {
-        data.utm_campaign = metadata.utmCampaign
-        data.utm_content = metadata.utmContent
-        data.utm_medium = metadata.utmMedium
-        data.utm_source = metadata.utmSource
-        data.utm_term = metadata.utmTerm
+var SCHEMA = isA.object({
+  flowId: isA.string().length(64).regex(HEX).optional(),
+  flowBeginTime: isA.number().integer().positive().optional(),
+  context: isA.string().optional(),
+  entrypoint: isA.string().optional(),
+  migration: isA.string().optional(),
+  service: isA.string().optional(),
+  utmCampaign: isA.string().optional(),
+  utmContent: isA.string().optional(),
+  utmMedium: isA.string().optional(),
+  utmSource: isA.string().optional(),
+  utmTerm: isA.string().optional()
+}).and('flowId', 'flowBeginTime').optional()
+
+var NOP = function () {
+  return P.resolve()
+}
+
+var NULL_MEMCACHED = {
+  delAsync: NOP,
+  getAsync: NOP,
+  setAsync: NOP
+}
+
+module.exports = function (log, config) {
+  var _memcached
+
+  return {
+    schema: SCHEMA,
+    save: save,
+    copy: copy,
+    remove: remove
+  }
+
+  /**
+   * Saves metrics context metadata against a token.
+   *
+   * @param token    token to save the metadata against
+   * @param events   array of event names that constitute a flow
+   * @param metadata metrics context metadata
+   */
+  function save (token, events, metadata) {
+    if (! token || ! events || ! metadata) {
+      return P.resolve()
+    }
+
+    if (typeof events === 'string') {
+      events = [ events ]
+    }
+
+    if (! Array.isArray(events)) {
+      return P.resolve()
+    }
```

Should we throw/log an error here rather than silently succeeding?